src/key.rs

```diff
@@ -461,73 +529,85 @@ impl RsaPrivateKey {
         self.precomputed.as_ref().map(|p| &p.dp)
     }
 
-    fn dq(&self) -> Option<&BigUint> {
+    fn dq(&self) -> Option<&BoxedUint> {
         self.precomputed.as_ref().map(|p| &p.dq)
     }
 
-    fn qinv(&self) -> Option<&BigInt> {
+    fn qinv(&self) -> Option<&BoxedMontyForm> {
         self.precomputed.as_ref().map(|p| &p.qinv)
     }
 
     fn crt_values(&self) -> Option<&[CrtValue]> {
-        /* for some reason the standard self.precomputed.as_ref().map() doesn't work */
-        if let Some(p) = &self.precomputed {
-            Some(p.crt_values.as_slice())
-        } else {
-            None
-        }
+        None
+    }
+
+    fn p_params(&self) -> Option<&BoxedMontyParams> {
+        self.precomputed.as_ref().map(|p| &p.p_params)
+    }
+
+    fn q_params(&self) -> Option<&BoxedMontyParams> {
+        self.precomputed.as_ref().map(|p| &p.q_params)
     }
 }
 
 /// Check that the public key is well formed and has an exponent within acceptable bounds.
 #[inline]
 pub fn check_public(public_key: &impl PublicKeyParts) -> Result<()> {
-    check_public_with_max_size(public_key, RsaPublicKey::MAX_SIZE)
+    check_public_with_max_size(public_key.n(), public_key.e(), RsaPublicKey::MAX_SIZE)
 }
 
 /// Check that the public key is well formed and has an exponent within acceptable bounds.
 #[inline]
-fn check_public_with_max_size(public_key: &impl PublicKeyParts, max_size: usize) -> Result<()> {
-    if public_key.n().bits() > max_size {
+fn check_public_with_max_size(n: &BoxedUint, e: &BoxedUint, max_size: usize) -> Result<()> {
+    if n.bits_precision() as usize > max_size {
         return Err(Error::ModulusTooLarge);
     }
 
-    let e = public_key
-        .e()
-        .to_u64()
-        .ok_or(Error::PublicExponentTooLarge)?;
-
-    if public_key.e() >= public_key.n() || public_key.n().is_even() {
+    if e >= n || n.is_even().into() || n.is_zero().into() {
         return Err(Error::InvalidModulus);
     }
 
-    if public_key.e().is_even() {
+    if e.is_even().into() {
         return Err(Error::InvalidExponent);
     }
 
-    if e < RsaPublicKey::MIN_PUB_EXPONENT {
+    if e < &BoxedUint::from(RsaPublicKey::MIN_PUB_EXPONENT) {
         return Err(Error::PublicExponentTooSmall);
     }
 
-    if e > RsaPublicKey::MAX_PUB_EXPONENT {
+    if e > &BoxedUint::from(RsaPublicKey::MAX_PUB_EXPONENT) {
         return Err(Error::PublicExponentTooLarge);
     }
 
     Ok(())
 }
 
+pub(crate) fn reduce(n: &BoxedUint, p: &BoxedMontyParams) -> BoxedMontyForm {
+    let bits_precision = p.modulus().bits_precision();
+    let modulus = p.modulus().as_nz_ref().clone();
+
+    let n = match n.bits_precision().cmp(&bits_precision) {
+        Ordering::Less => n.widen(bits_precision),
+        Ordering::Equal => n.clone(),
+        Ordering::Greater => n.shorten(bits_precision),
+    };
+
+    let n_reduced = n.rem_vartime(&modulus).widen(p.bits_precision());
+    BoxedMontyForm::new(n_reduced, p.clone())
+}
```

Comment:
Am I missing where this is actually used?

Perhaps it should be named `reduce_vartime` since it uses `rem_vartime`?

Reply:
renamed, and moved it ot where it is used